[4/n][Optimus][Auto-AC] Expose the config to skip the dynamo gaurds to avoid recompile

mengluy0125 opened the pull request:
Summary:
context: https://fb.workplace.com/groups/1075192433118967/permalink/1673720956599442/


Thanks Microve for raising the existing dynamo skip API in D75196435

The dynamic shape triggers recompilation, introducing compilation time increase, we expose config that users can skip the dynamo guards to avoid the recompile. Note that it may quantize unnessarily nodes, which can impact NE, QPS and memory saving,  needs verification.

Differential Revision: D75248430




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1314,6 +1315,29 @@ def statically_known_true(x: BoolLikeType) -> bool:
 
+def statically_known_false(x: BoolLikeType) -> bool:
+    """
+    Returns True if x can be simplified to a constant and is true.
```
and is False

bobrenjc93 commented on torch/_functorch/partitioners.py:
```diff
@@ -491,10 +492,19 @@ def should_quantize(node: torch.fx.Node) -> bool:
+            "activation_quantization_aten_pass"
+        ].get("size_in_mb", 100)
+    else:
+        return statically_known_false(
```
Let me take some time to explain the motivation behind statically_known_{true,false}:

The idea behind these APIs is if we can deduce that this is expression is True given the current state of our environment (considering all guards and runtime assertions up to this point in time). This is very different from your proposed implementation, but for the sake of discussion let's assume you implemented like done in https://github.com/pytorch/pytorch/pull/154291

I propose rewriting this to the following:

```
threshold = torch._inductor.config.post_grad_fusion_options["activation_quantization_aten_pass"].get("size_in_mb", 100)
return statically_known_true(size_in_mb >= threshold) or not statically_known_false(size_in_mb >= threshold)
````

Now let's consider a few cases

Case1: static size_in_mb >= threshold
Here `statically_known_true(size_in_mb >= threshold)` returns true and correctly quantize

Case2: static size_in_mb < threshold
Here `statically_known_true(size_in_mb >= threshold)` returns false and `not statically_known_false(size_in_mb >= threshold)` returns false and we correctly don't quantize

Case3: dynamic size_in_mb >= treshold
Here `statically_known_true(size_in_mb >= threshold)` returns false and `not statically_known_false(size_in_mb >= threshold)` returns true and we correctly quantize

Case4: dynamic size_in_mb < threshold
Here `statically_known_true(size_in_mb >= threshold)` returns false and `not statically_known_false(size_in_mb >= threshold)` returns true and we correctly quantize

The reason why statically_known_{true,false} always returns False in the case of dynamic is because we truly don't know. The value range of size_in_mb could vary based on the runtime size. If the runtime size is 2, we would be under the threshold, if the runtime size is 10000 we would be over.

bobrenjc93 commented on torch/_functorch/partitioners.py:
```diff
@@ -491,10 +493,24 @@ def should_quantize(node: torch.fx.Node) -> bool:
     size_in_mb = calculate_tensor_size(node.meta["val"])
-
-    return size_in_mb >= torch._inductor.config.post_grad_fusion_options[
+    if torch._inductor.config.post_grad_fusion_options[
```
readability nit: would be easier if you factored out the threshold into a separate variable. especially since you use it in multiple places, might as well CSE here

mengluy0125 committed:
[4/n][Optimus][Auto-AC] Expose the config to skip the dynamo gaurds to avoid recompile (#154152)

Summary:

context: https://fb.workplace.com/groups/1075192433118967/permalink/1673720956599442/


Thanks Microve for raising the existing dynamo skip API in D75196435
Thanks bobrenjc93 for putting up a PR for statically_known_false

The dynamic shape triggers recompilation, introducing compilation time increase, we expose config that users can skip the dynamo guards to avoid the recompile. Note that it may quantize unnessarily nodes, which can impact NE, QPS and memory saving,  needs verification.

Test Plan:
### E2E (128 LWB omniFM)

| -- |
| Tech  |mast job | qps(k)  | compile time(s) |  peak memory(%)  | tlparse|
|  baseline | aps-ruilinchen-e76be3b96b  |  165.5 | 454  | 72.31 | [link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-ruilinchen-e76be3b96b/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000)|
| not skip guards  | aps-lwb_0514_fp8_quant-47d59ec171  | 169.1  | 1107  | 69.37 |  [link](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-lwb_0514_fp8_quant-47d59ec171/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000) |
| skip guards  |  aps-lwb_0514_fp8_quant_skip_dynamo_guards-d60384b685 |  164.5 |  697 | 69.02 |[link ](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-lwb_0514_fp8_quant_skip_dynamo_guards-d60384b685/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000)|

 {F1978532271}

- Without skipping dynamo,
-2.2% qps gain
-3% peak memory reduction
-2.44x compilation time

- With skipping dynamo
-0.6% qps drop
-3.3% peak memory reduction
-1.535x compilation time

Reviewed By: bobrenjc93

Differential Revision: D75248430

izaitsevfb commented:
@pytorchbot merge -f 'unrelated cancellation'
